# python functions(Assignment_9)

### 1. What is a lambda function in Python, and how does it differ from a regular function?

In Python, a lambda function is a small, anonymous function that can be defined inline without a formal name. It is also known as an "anonymous function" or a "function literal." Lambda functions are defined using the `lambda` keyword, followed by a list of arguments, a colon (`:`), and an expression that represents the function's body.

The syntax for a lambda function is as follows:
#### lambda arguments: expression


Here's an example that demonstrates the usage of a lambda function:


In [3]:
double = lambda x: x * 2
print(double(5))  # Output: 10

10


In this example, the lambda function takes an argument `x` and returns its doubled value.

The key differences between lambda functions and regular functions in Python are:

1. **Anonymous nature:** Lambda functions are anonymous, meaning they don't have a name associated with them. They are primarily used when you need a simple, one-line function and don't want to define a separate named function.

2. **Syntax and simplicity:** Lambda functions have a more concise syntax compared to regular functions. They consist of a single expression rather than a block of statements. This makes them useful for situations where a small, inline function is required.

3. **Limited functionality:** Lambda functions are limited in functionality compared to regular functions. They can only contain a single expression, which means they can't include complex logic or multiple statements. Regular functions, on the other hand, can have multiple statements, local variables, and more complex control flow.

4. **Usage as function arguments:** Lambda functions are commonly used as arguments for higher-order functions like `map()`, `filter()`, and `reduce()`. These functions accept other functions as parameters, and lambda functions provide a convenient way to define these functions inline without explicitly creating named functions.

It's worth noting that lambda functions are not a replacement for regular functions in Python. They are typically used in situations where a simple, one-line function is needed, such as in functional programming paradigms or when working with higher-order functions. For more complex or reusable functions, it is generally recommended to use regular named functions.

### 2. Can a lambda function in Python have multiple arguments? If yes, how can you define and use them?

Yes, a lambda function in Python can have multiple arguments. The syntax for defining a lambda function with multiple arguments is similar to that of a regular function.

The general syntax for a lambda function with multiple arguments is as follows:
#### lambda arg1, arg2, ..., argN: expression


Here's an example that demonstrates the usage of a lambda function with multiple arguments:

In [6]:
multiply = lambda x, y: x * y
print(multiply(3, 4))  # Output: 12

12


In this example, the lambda function takes two arguments, `x` and `y`, and returns their multiplication.

You can define lambda functions with any number of arguments by separating them with commas in the argument list. Just like regular functions, lambda functions can have zero or more arguments.

To use a lambda function with multiple arguments, you simply call it with the appropriate number of arguments, separated by commas, just like you would with a regular function.

Here's another example that uses a lambda function with three arguments:

In [8]:
add_three_numbers = lambda a, b, c: a + b + c
print(add_three_numbers(1, 2, 3))  # Output: 6

6


In this example, the lambda function `add_three_numbers` takes three arguments, `a`, `b`, and `c`, and returns their sum.

Remember that lambda functions are primarily used for simple, one-line functions. If you find yourself needing more complex logic or multiple statements, it is generally recommended to use regular named functions instead.

### 3. How are lambda functions typically used in Python? Provide an example use case.

Lambda functions in Python are commonly used in situations where a small, inline function is needed, especially when working with higher-order functions or functional programming paradigms. Here are a few example use cases for lambda functions:

1. **Mapping:** Lambda functions are frequently used with the `map()` function to transform elements of an iterable. The `map()` function takes a function and an iterable as arguments, and it applies the function to each element of the iterable, returning an iterator of the results. Lambda functions are convenient for defining the transformation logic inline. For example:

In [10]:
numbers = [1, 2, 3, 4, 5]
squared = list(map(lambda x: x**2, numbers))
print(squared)

[1, 4, 9, 16, 25]


In this example, the lambda function squares each number in the `numbers` list using `map()`, resulting in a new list of squared numbers.

2. **Filtering:** Lambda functions are also commonly used with the `filter()` function to select elements from an iterable based on a condition. The `filter()` function takes a function that returns a Boolean value and an iterable, and it returns an iterator containing the elements for which the function returns `True`. Lambda functions can be used to define the filtering condition inline. For example:

In [12]:
numbers = [1, 2, 3, 4, 5]
even_numbers = list(filter(lambda x: x % 2 == 0, numbers))
print(even_numbers)

[2, 4]


In this example, the lambda function filters out the odd numbers from the `numbers` list using `filter()`, resulting in a new list of even numbers.

3. **Sorting:** Lambda functions are useful for defining custom sorting criteria when working with functions like `sorted()` or `sort()`. These functions take a `key` parameter that specifies a function to extract a comparison key from each element. Lambda functions can be used to define the key extraction logic inline. For example:

In [14]:
people = [('Alice', 25), ('Bob', 32), ('Charlie', 18)]
people_sorted_by_age = sorted(people, key=lambda person: person[1])
print(people_sorted_by_age)

[('Charlie', 18), ('Alice', 25), ('Bob', 32)]




In this example, the lambda function extracts the second element (age) from each tuple in the `people` list to determine the sorting order based on age.

These are just a few examples of how lambda functions can be used in Python. Lambda functions provide a concise and convenient way to define small, disposable functions on the fly, making code more expressive and compact.

### 4. What are the advantages and limitations of lambda functions compared to regular functions in Python?

Lambda functions in Python have certain advantages and limitations compared to regular functions. Let's examine them:

Advantages of lambda functions:

1. **Concise syntax**: Lambda functions have a more compact syntax compared to regular functions. They allow you to define simple, one-line functions without the need for a formal function declaration or a separate name.

2. **Inline definition**: Lambda functions are defined inline, meaning they can be directly embedded within the code where they are used. This eliminates the need to define a separate function elsewhere in the code, making the code more streamlined and focused.

3. **Functional programming**: Lambda functions are particularly useful when working with functional programming concepts such as higher-order functions, mapping, filtering, and reducing. They enable you to express transformations and operations on data in a more expressive and concise manner.

Limitations of lambda functions:

1. **Limited functionality**: Lambda functions are limited in their functionality compared to regular functions. They are restricted to a single expression and cannot include multiple statements, local variables, or complex control flow. This limitation makes them suitable for simple computations but less suitable for more complex tasks.

2. **Reduced readability**: While lambda functions can be concise, their brevity can sometimes lead to reduced readability, especially if the expression becomes overly complex or convoluted. Regular functions with descriptive names and well-defined blocks of code can be more readable and easier to understand, particularly for larger or reusable functions.

3. **Debugging and traceability**: Lambda functions lack a named identifier, which can make debugging more challenging. When an error occurs within a lambda function, the traceback will not provide a specific function name, making it harder to identify the source of the error.

4. **Reusability**: Lambda functions are primarily intended for short, disposable functions. They are not easily reusable or shareable across different parts of the codebase. Regular functions, on the other hand, can be defined once and used in multiple places, promoting code reusability and maintainability.

In summary, lambda functions offer conciseness and convenience for simple, one-line functions, especially when working with functional programming concepts. However, they have limitations in terms of functionality, readability, debuggability, and reusability, which may make regular functions a more suitable choice for more complex or reusable code.

### 5. Are lambda functions in Python able to access variables defined outside of their own scope? Explain with an example.

Yes, lambda functions in Python have access to variables defined outside of their own scope. This concept is known as "lexical scoping" or "closure."

When a lambda function is defined, it can access variables from the surrounding scope in which it is defined. This includes variables from the global scope or variables defined in an enclosing function. The lambda function "closes over" these variables and retains access to them even if they go out of scope or are modified later.

Here's an example to illustrate this behavior:

In [18]:
def generate_multiplier(factor):
    return lambda x: x * factor

multiply_by_2 = generate_multiplier(2)
multiply_by_3 = generate_multiplier(3)

print(multiply_by_2(5))  
print(multiply_by_3(5))  

10
15


In this example, the `generate_multiplier()` function takes a `factor` parameter and returns a lambda function that multiplies a given value by the `factor`. We create two instances of the lambda function: `multiply_by_2` and `multiply_by_3`. Each instance "captures" the value of `factor` from its respective invocation of `generate_multiplier()`.

Even though the `factor` variable is not directly passed as an argument to the lambda functions, they still retain access to it because they are defined within the lexical scope of the `generate_multiplier()` function. This allows the lambda functions to multiply values by the corresponding factor even after the `generate_multiplier()` function has completed its execution.

It's important to note that lambda functions do not create a separate copy of the variables they access. Instead, they maintain references to the variables in the enclosing scope. Therefore, if the variable is mutable (e.g., a list or dictionary), modifications made to it will be reflected in the lambda function as well.

This ability to access variables from the enclosing scope makes lambda functions powerful for creating closures and implementing functional programming techniques.

### 6. Write a lambda function to calculate the square of a given number

Here's a lambda function that calculates the square of a given number:

In [21]:
square = lambda x: x ** 2

In this lambda function, the argument `x` is squared using the `**` operator, and the result is returned as the output.

You can use this lambda function to calculate the square of a number by calling it with the desired value. Here's an example:

In [22]:
result = square(5)
print(result) 

25


In this example, we call the `square` lambda function with the argument `5`, which computes the square of `5` and assigns the result to the variable `result`. Finally, we print the value of `result`, which outputs `25` as the square of `5`.

### 7. Create a lambda function to find the maximum value in a list of integers.

Here's a lambda function that finds the maximum value in a list of integers:

In [28]:
find_max = lambda lst: max(lst)

In this lambda function, the argument `lst` represents the list of integers. The `max()` function is used to find the maximum value within the list, and that value is returned as the output of the lambda function.

You can use this lambda function to find the maximum value in a list by calling it with the desired list as the argument. Here's an example:

In [29]:
numbers = [5, 8, 2, 10, 3]
result = find_max(numbers)
print(result)  

10



In this example, we have a list of integers called `numbers`. We call the `find_max` lambda function with `numbers` as the argument, which finds the maximum value within the list. The result is then assigned to the variable `result`, and we print its value, which outputs `10` as the maximum value in the list.

Note that the lambda function simply wraps the `max()` function. Its purpose is to provide a concise and inline way to find the maximum value without the need for a separate named function declaration.

### 8. Implement a lambda function to filter out all the even numbers from a list of integers.

Here's a lambda function that filters out all the even numbers from a list of integers:

In [50]:
filter_even = lambda lst: list(filter(lambda x: x % 2 == 0, lst))

In this lambda function, the argument `lst` represents the list of integers. The `filter()` function is used to apply a filtering condition to each element of the list. The inner lambda function checks if each element `x` is divisible by `2` (i.e., it is even) using the modulo operator `%`. Only the elements that satisfy the condition are included in the filtered result. The `list()` function is used to convert the filter object to a list.

You can use this lambda function to filter out even numbers from a list by calling it with the desired list as the argument. Here's an example:

In [51]:
numbers = [5, 8, 2, 10, 3]
result = filter_even(numbers)
print(result) 

[8, 2, 10]


In this example, we have a list of integers called `numbers`. We call the `filter_even` lambda function with `numbers` as the argument, which filters out the even numbers from the list. The result is then assigned to the variable `result`, and we print its value, which outputs `[8, 2, 10]` as the filtered list containing only the even numbers.

The lambda function provides a concise and inline way to filter elements based on a condition without the need for a separate named function declaration.

### 9. Write a lambda function to sort a list of strings in ascending order based on the length of each string.

Here's a lambda function that sorts a list of strings in ascending order based on the length of each string:

In [54]:
sort_by_length = lambda lst: sorted(lst, key=lambda s: len(s))

In this lambda function, the argument `lst` represents the list of strings. The `sorted()` function is used to perform the sorting operation. The `key` parameter is set to another lambda function, which extracts the length of each string `s` using the `len()` function. This lambda function acts as a key function, providing the basis for the sorting comparison.

You can use this lambda function to sort a list of strings by length in ascending order by calling it with the desired list as the argument. Here's an example:

In [56]:
strings = ['apple', 'banana', 'cherry', 'date', 'elderberry']
result = sort_by_length(strings)
print(result) 

['date', 'apple', 'banana', 'cherry', 'elderberry']



In this example, we have a list of strings called `strings`. We call the `sort_by_length` lambda function with `strings` as the argument, which sorts the strings in ascending order based on their lengths. The result is then assigned to the variable `result`, and we print its value, which outputs `['date', 'apple', 'banana', 'cherry', 'elderberry']` as the sorted list of strings.

The lambda function provides a concise and inline way to define the key function for sorting, allowing you to sort elements based on a specific criterion without the need for a separate named function declaration.

### 10. Create a lambda function that takes two lists as input and returns a new list containing the common elements between the two lists.

Here's a lambda function that takes two lists as input and returns a new list containing the common elements between the two lists:

In [2]:
find_common_elements = lambda list1, list2: list(filter(lambda x: x in list1, list2))

In this lambda function, the arguments `list1` and `list2` represent the two input lists. The `filter()` function is used to apply a filtering condition to each element of `list2`. The inner lambda function checks if each element `x` is present in `list1` using the `in` operator. Only the elements that satisfy the condition (i.e., they are common to both lists) are included in the filtered result. The `list()` function is used to convert the filter object to a list.

You can use this lambda function to find the common elements between two lists by calling it with the desired lists as arguments. Here's an example:

In [3]:
list1 = [1, 2, 3, 4, 5]
list2 = [4, 5, 6, 7, 8]
result = find_common_elements(list1, list2)
print(result) 

[4, 5]


In this example, we have two lists, `list1` and `list2`. We call the `find_common_elements` lambda function with `list1` and `list2` as arguments, which filters out the common elements between the two lists. The result is then assigned to the variable `result`, and we print its value, which outputs `[4, 5]` as the list containing the common elements between `list1` and `list2`.

The lambda function provides a concise and inline way to filter elements based on a condition without the need for a separate named function declaration.

### 11. Write a recursive function to calculate the factorial of a given positive integer.

Here's an example of a recursive function to calculate the factorial of a given positive integer:

In [14]:
def factorial(n):
    if n == 0:
        return 1
    else:
        return n * factorial(n - 1)

In this recursive function, the base case is when `n` is equal to 0. In that case, the function returns 1, as the factorial of 0 is defined as 1. For any other positive integer `n`, the function calls itself with the argument `n - 1` and multiplies the result by `n`.

You can use this recursive function to calculate the factorial of a positive integer by calling it with the desired value. Here's an example:

In [15]:
result = factorial(5)
print(result)  

120



In this example, we call the `factorial` function with the argument `5`, which calculates the factorial of 5 using recursion. The result, which is `120`, is assigned to the variable `result`, and we print its value.

The recursive function calculates the factorial by breaking down the problem into smaller subproblems until it reaches the base case. It's important to ensure that the function is called with a positive integer to avoid infinite recursion.

### 12. Implement a recursive function to compute the nth Fibonacci number.

Here's an example of a recursive function to compute the nth Fibonacci number:

In [18]:
def fibonacci(n):
    if n <= 1:
        return n
    else:
        return fibonacci(n - 1) + fibonacci(n - 2)

In this recursive function, the base case is when `n` is less than or equal to 1. In that case, the function returns `n` itself, as the Fibonacci sequence starts with 0 and 1. For any other positive integer `n`, the function calls itself recursively with the arguments `n - 1` and `n - 2`, and returns the sum of the two previous Fibonacci numbers.

You can use this recursive function to compute the nth Fibonacci number by calling it with the desired value. Here's an example:

In [19]:
result = fibonacci(6)
print(result)  

8


In this example, we call the `fibonacci` function with the argument `6`, which computes the 6th Fibonacci number using recursion. The result, which is `8`, is assigned to the variable `result`, and we print its value.

The recursive function calculates the Fibonacci sequence by breaking down the problem into smaller subproblems until it reaches the base case. It's important to note that the recursive approach can be computationally expensive for large values of `n`, as it involves redundant calculations. Memoization or dynamic programming techniques can be applied to optimize the performance of the Fibonacci calculation.

### 13. Create a recursive function to find the sum of all the elements in a given list.

Here's an example of a recursive function to find the sum of all the elements in a given list:

In [24]:
def list_sum(lst):
    if len(lst) == 0:
        return 0
    else:
        return lst[0] + list_sum(lst[1:])

In this recursive function, the base case is when the length of the list `lst` is equal to 0. In that case, the function returns 0, as the sum of an empty list is 0. For any non-empty list, the function recursively adds the first element `lst[0]` to the sum of the remaining elements `list_sum(lst[1:])`.

You can use this recursive function to find the sum of all the elements in a given list by calling it with the desired list as the argument. Here's an example:

In [25]:
numbers = [1, 2, 3, 4, 5]
result = list_sum(numbers)
print(result)  


15


In this example, we have a list of numbers called `numbers`. We call the `list_sum` function with `numbers` as the argument, which computes the sum of all the elements in the list using recursion. The result, which is `15`, is assigned to the variable `result`, and we print its value.

The recursive function calculates the sum by breaking down the problem into smaller subproblems until it reaches the base case. It's important to ensure that the function is called with a non-empty list to avoid any errors.

### 14. Write a recursive function to determine whether a given string is a palindrome.

Here's an example of a recursive function to determine whether a given string is a palindrome:

In [34]:
def is_palindrome(s):
    if len(s) <= 1:
        return True
    else:
        if s[0] == s[-1]:
            return is_palindrome(s[1:-1])
        else:
            return False

In this recursive function, the base case is when the length of the string `s` is less than or equal to 1. In that case, the function returns `True`, as a string with only one character or no characters is considered a palindrome. For any other string, the function checks if the first and last characters of the string are equal. If they are, the function calls itself recursively with the substring `s[1:-1]`, excluding the first and last characters. If the first and last characters are not equal, the function returns `False`.

You can use this recursive function to determine whether a given string is a palindrome by calling it with the desired string as the argument. Here's an example:

In [35]:
string1 = "radar"
string2 = "hello"
result1 = is_palindrome(string1)
result2 = is_palindrome(string2)
print(result1) 
print(result2)  

True
False


In this example, we have two strings, `string1` and `string2`. We call the `is_palindrome` function with `string1` and `string2` as arguments, which determines whether they are palindromes using recursion. The results, which are `True` for `string1` and `False` for `string2`, are assigned to the variables `result1` and `result2`, respectively. We print their values.

The recursive function checks whether the given string is a palindrome by comparing the first and last characters, and then recursively checks the remaining substring. It continues this process until it reaches the base case or finds a mismatched pair of characters.

### 15. Implement a recursive function to find the greatest common divisor (GCD) of two positive integers.

Here's an example of a recursive function to find the greatest common divisor (GCD) of two positive integers:

In [38]:
def gcd(a, b):
    if b == 0:
        return a
    else:
        return gcd(b, a % b)

In this recursive function, the base case is when the second integer `b` is equal to 0. In that case, the function returns the first integer `a`, as the GCD of `a` and 0 is `a`. For any other positive integers `a` and `b`, the function calls itself recursively with the arguments `b` and `a % b`, which computes the remainder of `a` divided by `b`. This process continues until the base case is reached and the GCD is found.

You can use this recursive function to find the GCD of two positive integers by calling it with the desired values. Here's an example:

In [40]:
result = gcd(24, 36)
print(result)  

12



In this example, we call the `gcd` function with the arguments `24` and `36`, which computes the GCD of 24 and 36 using recursion. The result, which is `12`, is assigned to the variable `result`, and we print its value.

The recursive function calculates the GCD by using the Euclidean algorithm, which repeatedly subtracts the smaller number from the larger number until the remainder is 0. The GCD is then the remaining nonzero number. The recursive approach simplifies the computation by repeatedly reducing the problem to a smaller subproblem until the base case is reached.